### ISSUE TITLE
8.2.0 update breaks Python type hint support in _ren.py files and python blocks

### ISSUE THREAD

**Issue by micka190** [renpy/renpy#5316]
I've been using the new `_ren.py` feature to get better Python tooling support, including type hints.

The new 8.2.0 seems to break files that include Python type hints (i.e. `def foo(bar: str) -> bool:`).

Note that this affects both `_ren.py` files and `python:` code blocks in `.rpy` files (though I'm not sure if those were supposed to support Python type hints or not).

Edit: Updated the title twice, because I guess Ren'Py was using some kind of caching. It used to only happen when building a distribution or doing the "update `old_game`" command, but now I can confirm that it happens when trying to run the game in a development environment as well, so it seems the update broke type hints across the board.

**Comment by Gouvernathor** [comment]
Read the [incompatible changes](https://www.renpy.org/dev-doc/html/incompatible.html#incompatible-7-7-0). Type hints like the one you showed, using builtin types, should still do fine, the problem is with the use of the cancelled syntax enabled by `from __future__ import annotations` which (wrongly) enabled you to hint types that didn't exist at the time the hinted code is executed. It _seems_ normal.

**Comment by micka190** [comment]
*Facepalm*

When I read "**Stringified** annotations", I thought Ren'Py was doing something related to serialization or something. My bad.

In case anyone else has this issue with `_ren.py` files specifically, you can find and replace:

```python
"""renpy
init python:
"""
```

With this:

```python
"""renpy
rpy python annotations
init python:
"""
```

Kind of wish there was a way to enable it globally, to be honest, but I just updated the code snippet I use in VS Code that generates the renpy block to include it.

Since anything above it will be excluded from the file, and placing it below it won't work since it won't be valid Python code.

___


As a follow-up question, in my case it was probably because I was importing things like `Optional`, `List`, `Set`, etc. from the `typing` module *above* the `"""renpy` comment block (so it doesn't get included in the game, because I just wanted the tooling support).

Is there a "best practice" for where to include imports from the `typing` module when dealing with `_ren.py` files? Are they safe to leave above the `"""renpy` block as long as we have `rpy python annotations` or should I *not* have the `rpy python annotations` line and place the `typing` imports *below* the `"""renpy" block?

**Comment by Gouvernathor** [comment]
I would suggest either using rpy python annotations inside the renpy""" block, or importing typing. The former will work as long as Python supports `from __future__ import annotations`, and the latter will work as long as you don't use types and features only present in the latest python versions.
Ultimately all this will be resolved (hopefully) in Python 3.13, but it is not yet released and the date when renpy will move to that version is even more subject to speculations.

In the end, is there any specific difficulty between this and the old-game feature ?

**Comment by micka190** [comment]
> I would suggest either using rpy python annotations inside the renpy""" block, or importing typing.

Alright, I 'll keep it in the `"""renpy` block until it's baked into Python 3 itself (and hopefully Ren'Py).

> In the end, is there any specific difficulty between this and the old-game feature ?

No, I don't think so. I noticed the issue when I tried to make a build with the new version to make sure everything still worked in my project, but I got errors at the `update_old_game` step. I think it's because my build environment tried to recompile the files into `.rpyc` files, whereas the dev environment didn't because I hadn't made code changes yet. Once I changed files I started seeing the issue with dev builds too.

Gonna close the issue since it seems like I just misunderstood the incompatibilities introduced by the update.
